crates/wasmtime/src/component/func/host.rs

```diff
@@ -61,6 +62,24 @@ impl HostFunc {
         })
     }
 
+    pub(crate) fn new_dynamic<
+        T,
+        F: Fn(StoreContextMut<'_, T>, &[Val]) -> Result<Val> + Send + Sync + 'static,
+    >(
+        func: F,
+        params: Box<[Type]>,
+        result: Type,
+    ) -> Arc<HostFunc> {
+        Arc::new(HostFunc {
+            entrypoint: dynamic_entrypoint::<T, F>,
+            typecheck: typecheck_no_op,
```

Comment:
Eventually I think this will need to be filled out when the function type is an argument to `func_new`. Additionally for now I think it would be good to go ahead and fill out to have an "extra layer" of defense against bugs.

For now I think the `new_dynamic` here could take a `TypeFuncIndex` and the implementation could basically test that the two function indices are the same (since everything comes from within the same component right now)